## 1. Import Libraries

In [381]:
import pandas as pd

## 2. Load Datasets

In [382]:
categories_df = pd.read_json("IN_category_id.json")
categories_df = pd.json_normalize(categories_df["items"])

df = pd.read_csv("In_youtube_trending_data.csv")

## 3. Initial Data Inspection

### a) Youtube Trending Dataset

In [383]:
df.head()

,video_id,title,publishedAt,channelId,channelTitle,categoryId,trending_date,tags,view_count,likes,dislikes,comment_count,thumbnail_link,comments_disabled,ratings_disabled,description
0,Iot0eF6EoNA,Sadak 2 | Official Trailer | Sanjay | Pooja | ...,2020-08-12T04:31:41Z,UCGqvJPRcv7aVFun-eTsatcA,FoxStarHindi,24,2020-08-12T00:00:00Z,sadak|sadak 2|mahesh bhatt|vishesh films|pooja...,9885899,224925,3979409,350210,https://i.ytimg.com/vi/Iot0eF6EoNA/default.jpg,False,False,Three Streams. Three Stories. One Journey. Sta...
1,x-KbnJ9fvJc,Kya Baat Aa : Karan Aujla (Official Video) Tan...,2020-08-11T09:00:11Z,UCm9SZAl03Rev9sFwloCdz1g,Rehaan Records,10,2020-08-12T00:00:00Z,[None],11308046,655450,33242,405146,https://i.ytimg.com/vi/x-KbnJ9fvJc/default.jpg,False,False,Singer/Lyrics: Karan Aujla Feat Tania Music/ D...
2,KX06ksuS6Xo,Diljit Dosanjh: CLASH (Official) Music Video |...,2020-08-11T07:30:02Z,UCZRdNleCgW-BGUJf-bbjzQg,Diljit Dosanjh,10,2020-08-12T00:00:00Z,clash diljit dosanjh|diljit dosanjh|diljit dos...,9140911,296533,6179,30058,https://i.ytimg.com/vi/KX06ksuS6Xo/default.jpg,False,False,CLASH official music video performed by DILJIT...
3,UsMRgnTcchY,Dil Ko Maine Di Kasam Video | Amaal M Ft.Ariji...,2020-08-10T05:30:49Z,UCq-Fj5jknLsUf-MWSy4_brA,T-Series,10,2020-08-12T00:00:00Z,hindi songs|2020 hindi songs|2020 new songs|t-...,23564512,743931,84162,136942,https://i.ytimg.com/vi/UsMRgnTcchY/default.jpg,False,False,Gulshan Kumar and T-Series presents Bhushan Ku...
4,WNSEXJJhKTU,"Baarish (Official Video) Payal Dev,Stebin Ben ...",2020-08-11T05:30:13Z,UCye6Oz0mg46S362LwARGVcA,VYRLOriginals,10,2020-08-12T00:00:00Z,VYRL Original|Mohsin Khan|Shivangi Joshi|Payal...,6783649,268817,8798,22984,https://i.ytimg.com/vi/WNSEXJJhKTU/default.jpg,False,False,VYRL Originals brings to you ‘Baarish’ - the b...


In [384]:
df.describe(include="all")

,video_id,title,publishedAt,channelId,channelTitle,categoryId,trending_date,tags,view_count,likes,dislikes,comment_count,thumbnail_link,comments_disabled,ratings_disabled,description
count,251277,251277,251277,251277,251276,251277.000000,251277,251277,2.512770e+05,2.512770e+05,2.512770e+05,2.512770e+05,251277,251277,251277,231822
unique,78847,79952,78160,4521,4802,NaN,1323,49787,NaN,NaN,NaN,NaN,78849,2,2,66830
top,y_RHQ4mBDKA,Kundali Bhagya,2021-06-04T18:15:01Z,UCvrhwpnp2DHYQ1CbXby9ypQ,Vijay Television,NaN,2021-03-02T00:00:00Z,[None],NaN,NaN,NaN,NaN,https://i.ytimg.com/vi/y_RHQ4mBDKA/default.jpg,False,False,Follow me on Instagram- https://www.instagram....
freq,19,55,20,3396,3396,NaN,398,43717,NaN,NaN,NaN,NaN,19,249580,248987,1652
mean,NaN,NaN,NaN,NaN,NaN,20.861448,NaN,NaN,2.793858e+06,1.389707e+05,2.333248e+03,8.246164e+03,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,6.040176,NaN,NaN,6.979626e+06,3.941973e+05,7.199925e+04,7.023638e+04,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,1.000000,NaN,NaN,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,20.000000,NaN,NaN,3.779310e+05,1.204900e+04,0.000000e+00,3.420000e+02,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,24.000000,NaN,NaN,9.474340e+05,3.699100e+04,0.000000e+00,1.123000e+03,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,24.000000,NaN,NaN,2.420593e+06,1.161690e+05,7.180000e+02,3.951000e+03,NaN,NaN,NaN,NaN


In [385]:
df.dtypes

video_id             object
title                object
publishedAt          object
channelId            object
channelTitle         object
categoryId            int64
trending_date        object
tags                 object
view_count            int64
likes                 int64
dislikes              int64
comment_count         int64
thumbnail_link       object
comments_disabled      bool
ratings_disabled       bool
description          object
dtype: object

In [386]:
df.shape

(251277, 16)

### b) Category Lookup Dataset

In [387]:
categories_df.head()

,kind,etag,id,snippet.title,snippet.assignable,snippet.channelId
0,youtube#videoCategory,IfWa37JGcqZs-jZeAyFGkbeh6bc,1,Film & Animation,True,UCBR8-60-B28hp2BmDPdntcQ
1,youtube#videoCategory,5XGylIs7zkjHh5940dsT5862m1Y,2,Autos & Vehicles,True,UCBR8-60-B28hp2BmDPdntcQ
2,youtube#videoCategory,HCjFMARbBeWjpm6PDfReCOMOZGA,10,Music,True,UCBR8-60-B28hp2BmDPdntcQ
3,youtube#videoCategory,ra8H7xyAfmE2FewsDabE3TUSq10,15,Pets & Animals,True,UCBR8-60-B28hp2BmDPdntcQ
4,youtube#videoCategory,7mqChSJogdF3hSIL-88BfDE-W8M,17,Sports,True,UCBR8-60-B28hp2BmDPdntcQ


In [388]:
categories_df.describe(include="all")

,kind,etag,id,snippet.title,snippet.assignable,snippet.channelId
count,31,31,31,31,31,31
unique,1,31,31,30,2,1
top,youtube#videoCategory,IfWa37JGcqZs-jZeAyFGkbeh6bc,1,Comedy,False,UCBR8-60-B28hp2BmDPdntcQ
freq,31,1,1,2,17,31


In [389]:
categories_df.dtypes

kind                  object
etag                  object
id                    object
snippet.title         object
snippet.assignable      bool
snippet.channelId     object
dtype: object

In [390]:
categories_df.shape

(31, 6)

## 4. Clean YouTube Trending Dataset

### 4.1 Drop Unnecessary Columns

In [391]:
df=df.drop(columns=["channelId","thumbnail_link","description"])

In [392]:
df.columns

Index(['video_id', 'title', 'publishedAt', 'channelTitle', 'categoryId',
       'trending_date', 'tags', 'view_count', 'likes', 'dislikes',
       'comment_count', 'comments_disabled', 'ratings_disabled'],
      dtype='object')

### 4.2 Fix Data Types

In [393]:
df["publishedAt"] = pd.to_datetime(df["publishedAt"])
df["trending_date"] = pd.to_datetime(df["trending_date"])

In [394]:
df.dtypes

video_id                          object
title                             object
publishedAt          datetime64[ns, UTC]
channelTitle                      object
categoryId                         int64
trending_date        datetime64[ns, UTC]
tags                              object
view_count                         int64
likes                              int64
dislikes                           int64
comment_count                      int64
comments_disabled                   bool
ratings_disabled                    bool
dtype: object

### 4.3 Handle Missing Values


In [395]:
df.isnull().sum()

video_id             0
title                0
publishedAt          0
channelTitle         1
categoryId           0
trending_date        0
tags                 0
view_count           0
likes                0
dislikes             0
comment_count        0
comments_disabled    0
ratings_disabled     0
dtype: int64

In [396]:
df = df.dropna()

### 4.4 Remove Duplicate Rows

In [397]:
df.duplicated().sum()

np.int64(75)

In [398]:
df = df.drop_duplicates()

In [399]:
df.shape

(251201, 13)

### 4.5 Removing Leading/Trailing spaces from Text Columns 

In [400]:
text_cols = ["title", "channelTitle"]

for col in text_cols:
    invalid = (df[col] != df[col].str.strip()).sum()
    print(f"{col}: {invalid}")

title: 44
channelTitle: 1491


In [401]:
df["title"] = df["title"].str.strip()
df["channelTitle"] = df["channelTitle"].str.strip()

### 4.6 Standardize Column Names

In [402]:
df = df.rename(columns={
    "title": "video_title",
    "publishedAt": "published_at",
    "channelTitle": "channel_title",
    "categoryId": "category_id",
})

### 4.7 Create Language Column

In [403]:
language_keywords = {
    "English" : ["english"],
    "Hindi": ["hindi"],
    "Tamil": ["tamil"],
    "Telugu": ["telugu"],
    "Malayalam": ["malayalam"],
    "Kannada": ["kannada"],
    "Punjabi": ["punjabi"],
    "Bengali": ["bengali", "bangla"],
    "Marathi": ["marathi"],
    "Gujarati": ["gujarati"],
    "Odia": ["odia", "oriya"]
}

def detect_language(row):
    title = str(row["video_title"]).lower()
    tags = str(row["tags"]).lower()

    for language, keywords in language_keywords.items():
        if any(keyword in title for keyword in keywords):
            return language
        elif any(keyword in tags for keyword in keywords):
            return language

    return "Unknown"

df["language"] = df.apply(detect_language, axis=1)

In [404]:
df["language"].value_counts(normalize=True)

language
Unknown      0.577211
Hindi        0.134824
Tamil        0.086222
Telugu       0.070047
Malayalam    0.050087
Punjabi      0.033881
English      0.009984
Kannada      0.009319
Odia         0.008873
Bengali      0.007114
Marathi      0.006353
Gujarati     0.006083
Name: proportion, dtype: float64

- 57.7% of videos are classified as Unknown, meaning more than half the dataset could not be mapped to a specific language. This is because language detection relies on keyword matching in video titles and tags — videos without explicit language keywords default to Unknown. 
- Language-based analysis should be interpreted with caution.

### 4.8 Drop Temporary Tags Column

In [405]:
df = df.drop(columns=["tags"])

### 4.9 Checking 0 Values in Numeric Columns 

In [406]:
numeric_cols = ["view_count", "likes", "dislikes", "comment_count"]

for col in numeric_cols:
    zero_count = (df[col] == 0).sum()
    print(f"{col}")
    print(f"Zero Values : {zero_count}")
    print(f"Percentage : {(zero_count/len(df))*100:.2f}%")
    print("-"*40)

view_count
Zero Values : 16
Percentage : 0.01%
----------------------------------------
likes
Zero Values : 2295
Percentage : 0.91%
----------------------------------------
dislikes
Zero Values : 158921
Percentage : 63.26%
----------------------------------------
comment_count
Zero Values : 2251
Percentage : 0.90%
----------------------------------------


In [407]:
df.loc[df["view_count"] == 0,[
        "video_title",
        "channel_title",
        "published_at",
        "trending_date",
        "likes",
        "dislikes",
        "comment_count"
    ]
]

,video_title,channel_title,published_at,trending_date,likes,dislikes,comment_count
46737,Earth Day 2021 Doodle,GoogleDoodles,2021-04-21 11:00:14+00:00,2021-04-22 00:00:00+00:00,0,0,0
46895,Earth Day 2021 Doodle,GoogleDoodles,2021-04-21 11:00:14+00:00,2021-04-23 00:00:00+00:00,0,0,0
47058,Earth Day 2021 Doodle,GoogleDoodles,2021-04-21 11:00:14+00:00,2021-04-24 00:00:00+00:00,0,0,0
47242,Earth Day 2021 Doodle,GoogleDoodles,2021-04-21 11:00:14+00:00,2021-04-25 00:00:00+00:00,0,0,0
47428,Earth Day 2021 Doodle,GoogleDoodles,2021-04-21 11:00:14+00:00,2021-04-26 00:00:00+00:00,0,0,0
47639,Earth Day 2021 Doodle,GoogleDoodles,2021-04-21 11:00:14+00:00,2021-04-27 00:00:00+00:00,0,0,0
47846,Earth Day 2021 Doodle,GoogleDoodles,2021-04-21 11:00:14+00:00,2021-04-28 00:00:00+00:00,0,0,0
74085,Tim Bergling's 32nd Birthday,GoogleDoodles,2021-09-07 12:00:08+00:00,2021-09-08 00:00:00+00:00,0,0,0
74296,Tim Bergling's 32nd Birthday,GoogleDoodles,2021-09-07 12:00:08+00:00,2021-09-09 00:00:00+00:00,0,0,0
74523,Tim Bergling's 32nd Birthday,GoogleDoodles,2021-09-07 12:00:08+00:00,2021-09-10 00:00:00+00:00,0,0,0


In [408]:
df = df[df["view_count"]!=0].reset_index(drop=True)

In [409]:
yt_update_date = pd.Timestamp("2021-11-10", tz="UTC")

post_update_videos = df[df["trending_date"] >= yt_update_date]
zero_dislikes_rows = (post_update_videos["dislikes"] == 0).sum()
total_rows = len(post_update_videos)

print(f"Post-update rows: {total_rows}")
print(f"Zero dislikes after cutoff: {zero_dislikes_rows}")
print(f"Percentage: {(zero_dislikes_rows/total_rows)*100:.2f}%")

Post-update rows: 164691
Zero dislikes after cutoff: 158319
Percentage: 96.13%


In [410]:
comments_check = df[df["comment_count"] == 0]["comments_disabled"].value_counts()
print(comments_check)

comments_disabled
True     1681
False     554
Name: count, dtype: int64


- **likes** — 0.91% zero values. No systematic cause identified. Likely low-engagement videos, though a data collection issue cannot be fully ruled out. Retained as valid.

- **dislikes** — 63.26% zero values. 96.13% of these occur in videos trending after November 10, 2021, when YouTube made dislike counts private globally. From this date, the YouTube Data API stopped returning dislike counts entirely, so all values recorded as 0 after this date reflect API unavailability, not actual zero dislikes. 
The remaining 3.87% zero dislikes occur in videos trending before November 10, 2021, when dislikes were still publicly visible. These may reflect genuine zero dislikes or data collection inconsistencies — the exact cause cannot be determined.

- **comment_count** — 0.90% zero values. 75.2% have comments_disabled = True, which is expected. The remaining 24.8% have comments enabled — likely low-engagement videos with no 
comments, though a data collection issue cannot be fully ruled out. Retained as valid.

### 4.10 Data Validation and Profiling

#### a) Data Quality Check

In [411]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 251185 entries, 0 to 251184
Data columns (total 13 columns):
 #   Column             Non-Null Count   Dtype              
---  ------             --------------   -----              
 0   video_id           251185 non-null  object             
 1   video_title        251185 non-null  object             
 2   published_at       251185 non-null  datetime64[ns, UTC]
 3   channel_title      251185 non-null  object             
 4   category_id        251185 non-null  int64              
 5   trending_date      251185 non-null  datetime64[ns, UTC]
 6   view_count         251185 non-null  int64              
 7   likes              251185 non-null  int64              
 8   dislikes           251185 non-null  int64              
 9   comment_count      251185 non-null  int64              
 10  comments_disabled  251185 non-null  bool               
 11  ratings_disabled   251185 non-null  bool               
 12  language           251185 non-

#### b) Statistical Summary 

In [412]:
df.describe(include="all")

,video_id,video_title,published_at,channel_title,category_id,trending_date,view_count,likes,dislikes,comment_count,comments_disabled,ratings_disabled,language
count,251185,251185,251185,251185,251185.00000,251185,2.511850e+05,2.511850e+05,2.511850e+05,2.511850e+05,251185,251185,251185
unique,78844,79949,NaN,4786,NaN,NaN,NaN,NaN,NaN,NaN,2,2,12
top,y_RHQ4mBDKA,Kundali Bhagya,NaN,Vijay Television,NaN,NaN,NaN,NaN,NaN,NaN,False,False,Unknown
freq,19,55,NaN,3394,NaN,NaN,NaN,NaN,NaN,NaN,249504,248913,144980
mean,NaN,NaN,2022-06-01 03:08:12.538065408+00:00,NaN,20.86277,2022-06-03 06:04:57.447698176+00:00,2.794493e+06,1.390077e+05,2.333677e+03,8.248440e+03,NaN,NaN,NaN
min,NaN,NaN,2020-07-23 11:30:37+00:00,NaN,1.00000,2020-08-12 00:00:00+00:00,1.268600e+04,0.000000e+00,0.000000e+00,0.000000e+00,NaN,NaN,NaN
25%,NaN,NaN,2021-07-08 11:54:44+00:00,NaN,20.00000,2021-07-11 00:00:00+00:00,3.780010e+05,1.205300e+04,0.000000e+00,3.420000e+02,NaN,NaN,NaN
50%,NaN,NaN,2022-05-25 19:19:10+00:00,NaN,24.00000,2022-05-28 00:00:00+00:00,9.476610e+05,3.700300e+04,0.000000e+00,1.124000e+03,NaN,NaN,NaN
75%,NaN,NaN,2023-04-22 13:17:17+00:00,NaN,24.00000,2023-04-25 00:00:00+00:00,2.421026e+06,1.162030e+05,7.180000e+02,3.952000e+03,NaN,NaN,NaN
max,NaN,NaN,2024-04-15 05:47:57+00:00,NaN,29.00000,2024-04-15 00:00:00+00:00,2.644074e+08,1.611524e+07,1.234147e+07,6.738565e+06,NaN,NaN,NaN


#### c) Date Consistency Check

In [413]:
invalid_date_rows = df[df["published_at"] > df["trending_date"]]

print("Number of invalid rows:", len(invalid_date_rows))
display(invalid_date_rows)

Number of invalid rows: 10558


,video_id,video_title,published_at,channel_title,category_id,trending_date,view_count,likes,dislikes,comment_count,comments_disabled,ratings_disabled,language
0,Iot0eF6EoNA,Sadak 2 | Official Trailer | Sanjay | Pooja | ...,2020-08-12 04:31:41+00:00,FoxStarHindi,24,2020-08-12 00:00:00+00:00,9885899,224925,3979409,350210,False,False,Hindi
5,Xma15IYe5tQ,SISTER DOES MY HAIRCUT,2020-08-12 06:56:05+00:00,Mythpat,20,2020-08-12 00:00:00+00:00,1699326,332553,4627,75819,False,False,Unknown
9,XM4Kd1gQ7Ek,Jatti Jaandi (Official Video) | Kambi Ft. Mahi...,2020-08-12 02:24:34+00:00,Gringo Entertainments,10,2020-08-12 00:00:00+00:00,1466612,97192,2276,3311,False,False,Hindi
35,iRlIUrcKJa8,7 AM | ETV Telugu News | 12th Aug 2020,2020-08-12 02:24:50+00:00,ETV Telangana,25,2020-08-12 00:00:00+00:00,213515,2500,263,90,False,False,Telugu
52,8fK60WACK7s,બુધવાર જન્માષ્ટમી સ્પેશિયલ-૨૦૨૦ |BUDHWAR JANMA...,2020-08-12 00:30:00+00:00,T-Series Gujarati,10,2020-08-12 00:00:00+00:00,191829,1191,180,50,False,False,Gujarati
...,...,...,...,...,...,...,...,...,...,...,...,...,...
251139,JfvuuzzDQEc,Rabb Se Hai Dua | Premiere Ep 454 Preview - Ap...,2024-04-15 05:23:59+00:00,ZEE5,24,2024-04-15 00:00:00+00:00,305768,5039,0,62,False,False,Hindi
251148,zgTEYXG7zBA,GOLU NE DE DIYA BAHUT BADA SURPRISE | Armaan M...,2024-04-15 04:30:17+00:00,Malik Vlogs,22,2024-04-15 00:00:00+00:00,547268,27163,0,1952,False,False,Unknown
251151,-J0N5ut2_p4,SURPRISE Visit to Subscribers House for Ghee D...,2024-04-15 04:16:34+00:00,Flying Beast,22,2024-04-15 00:00:00+00:00,609969,51721,0,1730,False,False,Unknown
251156,Cfwu9PI-EfU,Kundali Bhagya | Premiere Ep 1837 Preview - Ap...,2024-04-15 05:47:57+00:00,ZEE5,24,2024-04-15 00:00:00+00:00,275464,3718,0,19,False,False,Hindi


In [414]:
(df["trending_date"].dt.time == pd.Timestamp("00:00:00").time()).all()

np.True_

In [415]:
(df["published_at"].dt.date > df["trending_date"].dt.date).sum()

np.int64(0)

- The dataset contains 10,558 rows where published_at appears later than trending_date. However, this is a timestamp artifact, not a data error.

- trending_date always records midnight (00:00:00), confirmed for all 251,185 rows. published_at carries the actual upload time (e.g., 04:31:41). So a video published at 04:31:41 on August 12 appears "later" than trending_date of August 12 00:00:00 purely due to the time component.

- When compared at the date level only, zero rows have published_at date greater than trending_date date — confirming no genuine date-order violations exist in the dataset.

#### d) Engagement Metric Integrity Check

In [416]:
print("Views >= Likes :",(df["view_count"] >= df["likes"]).all())

print("Views >= Comments :",(df["view_count"] >= df["comment_count"]).all())

print("Views >= Dislikes :", (df["view_count"] >= df["dislikes"]).all())

Views >= Likes : True
Views >= Comments : True
Views >= Dislikes : True


- Verified that view_count is greater than or equal to likes, dislikes, and comment_count for all rows. This confirms logical consistency in the engagement metrics — a video cannot have more likes, dislikes, or comments than total views.

- All three checks passed across all 251,185 rows.

### 4.11 Collapse to One Row Per Video

In [417]:
new_df = df.groupby("video_id").agg(
    video_title = ("video_title", "last"),
    published_at = ("published_at", "first"),
    channel_title = ("channel_title", "first"),
    language = ("language", "first"),
    category_id = ("category_id", "first"),
    first_trending_date = ("trending_date", "min"),
    last_trending_date = ("trending_date", "max"),
    view_count = ("view_count", "max"),
    likes = ("likes", "max"),
    dislikes = ("dislikes", "max"),
    comment_count = ("comment_count", "max"),
    comments_disabled = ("comments_disabled", "first"),
    ratings_disabled = ("ratings_disabled", "first")
).reset_index()

In [418]:
new_df["days_to_trend"] = (new_df["first_trending_date"].dt.date - new_df["published_at"].dt.date).apply(lambda x : x.days) 

In [419]:
new_df["days_trending"] = (new_df["last_trending_date"] - new_df["first_trending_date"]).dt.days + 1

In [420]:
new_df.head()

,video_id,video_title,published_at,channel_title,language,category_id,first_trending_date,last_trending_date,view_count,likes,dislikes,comment_count,comments_disabled,ratings_disabled,days_to_trend,days_trending
0,--518euSO44,അപൂർവ്വങ്ങളിൽ അപൂർവ്വം എന്നൊക്കെ ശരിക്കും പറയാ...,2023-05-20 13:30:00+00:00,Street Food Kerala,Unknown,22,2023-05-22 00:00:00+00:00,2023-05-24 00:00:00+00:00,480540,5744,0,261,False,False,2,3
1,--9rRxqTdjQ,Am I Married??? All Secrets Revealed🔥🔥🔥,2022-01-18 07:22:33+00:00,Gaurav Chaudhary,Unknown,22,2022-01-19 00:00:00+00:00,2022-01-22 00:00:00+00:00,529139,57368,0,2323,False,False,1,4
2,--GS_EPuuMc,गणपती स्थापना संपूर्ण पूजा विधी। गुरुजींनी सां...,2021-09-08 08:31:14+00:00,Webdunia Marathi,Marathi,22,2021-09-10 00:00:00+00:00,2021-09-12 00:00:00+00:00,366925,3531,342,172,False,False,2,3
3,--L_wDHUaxY,THIS MONSTER IS SO SCARY !,2021-03-04 05:06:56+00:00,Techno Gamerz,Unknown,20,2021-03-04 00:00:00+00:00,2021-03-07 00:00:00+00:00,4434301,553020,7316,22839,False,False,0,4
4,--NjHa7Dsls,Itne din se vlogs kyu nahi aa rahe ( Reason be...,2024-03-23 09:30:00+00:00,Dimple Malhan Vlogs,Unknown,22,2024-03-24 00:00:00+00:00,2024-03-26 00:00:00+00:00,2741947,156598,0,3960,False,False,1,3



- The raw dataset contains multiple rows per video — one for each day it appeared on the trending page. The engagement metrics (view_count, likes, dislikes, comment_count) are cumulative, meaning each row includes all previous days' totals. Keeping multiple rows per video causes any aggregation to produce inflated, misleading results.

- To fix this, each video is consolidated into a single row using video_id as the unique identifier, retaining the final cumulative metric values. The grouping also captures:
  - `first_trending_date` — the first day the video appeared on trending
  - `last_trending_date` — the last day the video appeared on trending

- Two additional columns are then derived from the above:
  - `days_to_trend` — days between publish date and first trending date
  (0 means the video trended on the same day it was published)
  - `days_trending` — total days the video stayed on the trending page

- After collapsing, the dataset contains one row per unique video, making all downstream analysis accurate and consistent.

## 5. Clean Category Lookup Dataset

### 5.1 Drop Unnecessary Columns

In [421]:
categories_df = categories_df.drop(columns=["kind","etag","snippet.assignable","snippet.channelId"])

### 5.2 Standardize Column Names

In [422]:
categories_df = categories_df.rename(columns={"id":"category_id","snippet.title":"category_name"})

### 5.3 Fix Data Types

In [423]:
categories_df.dtypes

category_id      object
category_name    object
dtype: object

In [424]:
categories_df["category_id"] = categories_df["category_id"].astype("int64")

### 5.4 Validating the Categories Data

In [425]:
categories_df.describe(include="all")

,category_id,category_name
count,31.000000,31
unique,NaN,30
top,NaN,Comedy
freq,NaN,2
mean,27.516129,NaN
std,11.389677,NaN
min,1.000000,NaN
25%,20.500000,NaN
50%,28.000000,NaN
75%,36.500000,NaN


In [426]:
categories_df["category_id"].unique()

array([ 1,  2, 10, 15, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 30,
       31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44])

In [427]:
categories_df.isnull().sum()

category_id      0
category_name    0
dtype: int64

## 6. Merge Datasets

In [428]:
merged_df = new_df.merge(categories_df,on="category_id",how='left')

### 6.1 Post Merge Validation

In [429]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78844 entries, 0 to 78843
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype              
---  ------               --------------  -----              
 0   video_id             78844 non-null  object             
 1   video_title          78844 non-null  object             
 2   published_at         78844 non-null  datetime64[ns, UTC]
 3   channel_title        78844 non-null  object             
 4   language             78844 non-null  object             
 5   category_id          78844 non-null  int64              
 6   first_trending_date  78844 non-null  datetime64[ns, UTC]
 7   last_trending_date   78844 non-null  datetime64[ns, UTC]
 8   view_count           78844 non-null  int64              
 9   likes                78844 non-null  int64              
 10  dislikes             78844 non-null  int64              
 11  comment_count        78844 non-null  int64              
 12  comments_disabled 

### 6.2 Handling Missing Values

In [430]:
merged_df["category_name"] = merged_df["category_name"].fillna("Unknown Category")

### 6.3 Drop Temporary Category ID Column 

In [431]:
merged_df = merged_df.drop(columns=["category_id"])

### 6.4 Reorder Columns

In [432]:
merged_df = merged_df[[
    "video_id",
    "video_title",
    "channel_title",
    "category_name",
    "language",
    "published_at",
    "first_trending_date",
    "last_trending_date",
    "days_to_trend",
    "days_trending",
    "view_count",
    "likes",
    "dislikes",
    "comment_count",
    "comments_disabled",
    "ratings_disabled"
]]

## 7. Export Cleaned Dataset

In [433]:
merged_df.to_csv("cleaned_india_youtube_trending_data.csv",index=False)

## Summary
- Raw dataset: 251,277 rows, one row per video per trending day, covering July 2020 to April 2024 - Dropped irrelevant columns, fixed data types, removed 1 null row, 75 duplicates, and 16 zero view_count rows
- Stripped whitespace, standardized column names to snake_case - Created language column via keyword matching — 57.7% classified as Unknown due to keyword-based detection limitations
- Merged with category lookup to attach category names - Consolidated to one row per unique video (78,847 rows) using video_id — added first_trending_date, last_trending_date, days_to_trend, and days_trending columns 
- 63.26% zero dislikes primarily due to YouTube hiding dislike counts globally from November 10, 2021
- Final dataset is accurate, consistent, and ready for analysis